libs/entitlement/asm-provider/src/test/java/org/elasticsearch/entitlement/instrumentation/impl/InstrumenterTests.java

```diff
@@ -330,6 +384,40 @@ public void testInstrumenterWorksWithInstanceMethodsAndOverloads() throws Except
         assertThat(getTestEntitlementChecker().checkSomeInstanceMethodCallCount, is(1));
     }
 
+    public void testInstrumenterWorksWithConstructors() throws Exception {
+        var classToInstrument = ClassToInstrument.class;
+
+        Map<MethodKey, CheckerMethod> methods = Map.of(
+            new MethodKey(classToInstrument.getName().replace('.', '/'), "<init>", List.of()),
+            getCheckerMethod(MockCtorEntitlementChecker.class, "checkCtor", Class.class),
+            new MethodKey(classToInstrument.getName().replace('.', '/'), "<init>", List.of("I")),
+            getCheckerMethod(MockCtorEntitlementChecker.class, "checkCtor", Class.class, int.class)
+        );
+
+        var instrumenter = createInstrumenter(methods);
+
+        byte[] newBytecode = instrumenter.instrumentClassFile(classToInstrument).bytecodes();
+
+        if (logger.isTraceEnabled()) {
+            logger.trace("Bytecode after instrumentation:\n{}", bytecode2text(newBytecode));
+        }
+
+        Class<?> newClass = new TestLoader(Testable.class.getClassLoader()).defineClassFromBytes(
+            classToInstrument.getName() + "_NEW",
+            newBytecode
+        );
+
+        getTestEntitlementChecker().isActive = true;
+
+        var ex = assertThrows(InvocationTargetException.class, () -> newClass.getConstructor().newInstance());
+        assertThat(ex.getCause(), instanceOf(TestException.class));
+        var ex2 = assertThrows(InvocationTargetException.class, () -> newClass.getConstructor(int.class).newInstance(123));
+        assertThat(ex2.getCause(), instanceOf(TestException.class));
+
+        assertThat(getTestEntitlementChecker().checkCtorCallCount, is(1));
+        assertThat(getTestEntitlementChecker().checkCtorIntCallCount, is(1));
```

Comment:
(I'd go so far as to say that I think nontrivial assertions are something of an antipattern. One major value of unit tests is to validate your design, because a design that's hard to test is probably hard to use and hard to understand. Sophisticated assertions make it easier to test code that's hard to test, and therefore to ignore these design feedback signals.)

Reply:
Yeah I'm not a fan of `assertTrue` or `assertFalse`. Those guys are like `assertThat` but with unhelpful messages.

When these are applicable, they make the code easier to understand at a glance:
- `assertEquals`
- `assertSame`
- `assertNull`
- `assertThrows`